go/src/github.com/mesosphere/rendler/scheduler/main.go

```diff
@@ -1,281 +1,284 @@
 package main
 
 import (
-	"code.google.com/p/goprotobuf/proto"
 	"container/list"
 	"encoding/json"
 	"flag"
 	"fmt"
-	"github.com/mesosphere/mesos-go/mesos"
+	"github.com/golang/protobuf/proto"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/mesosutil"
+	sched "github.com/mesos/mesos-go/scheduler"
 	"github.com/mesosphere/rendler"
 	"log"
 	"os"
 	"os/signal"
-	"time"
 	"path/filepath"
+	"time"
 )
 
-const TASK_CPUS = 0.1
-const TASK_MEM = 32.0
-const SHUTDOWN_TIMEOUT = 30  // in seconds
-
-func main() {
-	crawlQueue := list.New()  // list of string
-	renderQueue := list.New() // list of string
+const (
+	TASK_CPUS        = 0.1
+	TASK_MEM         = 32.0
+	SHUTDOWN_TIMEOUT = time.Duration(30) * time.Second
+)
 
-	processedURLs := list.New() // list of string
-	crawlResults := list.New()  // list of CrawlEdge
-	renderResults := make(map[string]string)
+var (
+	crawlCommand  = "python crawl_executor.py"
+	renderCommand = "python render_executor.py"
+)
 
-	seedUrl := flag.String("seed", "http://mesosphere.io", "The first URL to crawl")
-	master := flag.String("master", "127.0.1.1:5050", "Location of leading Mesos master")
-	localMode := flag.Bool("local", true, "If true, saves rendered web pages on local disk")
-	// TODO(nnielsen): Add flag for artifacts.
+var (
+	DEFAULT_FILTER = &mesos.Filters{RefuseSeconds: proto.Float64(1)}
+)
 
-	flag.Parse()
+func maxTasksForOffer(offer *mesos.Offer) int {
+	// TODO(nnielsen): Parse offer resources.
+	count := 0
 
-	crawlQueue.PushBack(*seedUrl)
+	var cpus float64 = 0
+	var mem float64 = 0
 
-	tasksCreated := 0
-	tasksRunning := 0
-	shuttingDown := false
+	for _, resource := range offer.Resources {
+		if resource.GetName() == "cpus" {
+			cpus = *resource.GetScalar().Value
+		}
 
-	crawlCommand := "python crawl_executor.py"
-	renderCommand := "python render_executor.py"
+		if resource.GetName() == "mem" {
+			mem = *resource.GetScalar().Value
+		}
+	}
 
-	if *localMode {
-		renderCommand += " --local"
+	for cpus >= TASK_CPUS && mem >= TASK_MEM {
+		count++
+		cpus -= TASK_CPUS
+		mem -= TASK_MEM
 	}
 
+	return count
+}
+
+type RendlerScheduler struct {
+	tasksCreated int
+	tasksRunning int
+
+	crawlQueue    *list.List
+	renderQueue   *list.List
+	processedURLs *list.List
+	crawlResults  *list.List
+	renderResults map[string]string
+
+	crawlExecutor  *mesos.ExecutorInfo
+	renderExecutor *mesos.ExecutorInfo
+
+	// this channel is closed when the program receives an interrupt,
+	// signalling that the program should shut down.
+	shutdown chan interface{}
+	// this channel is closed when there are no more outstanding tasks,
+	// and that the scheduler does not intend to receive more tasks
+	done chan interface{}
```

Comment:
Why do you need to do that in two stages?
On Tue 7 Jul 2015 at 23:29 Derek Chiang (Enchi Jiang) <
notifications@github.com> wrote:

> In go/src/github.com/mesosphere/rendler/scheduler/main.go
> <https://github.com/mesosphere/RENDLER/pull/29#discussion_r34116903>:
>
> > +
> > +	crawlQueue    *list.List
> > +	renderQueue   *list.List
> > +	processedURLs *list.List
> > +	crawlResults  *list.List
> > +	renderResults map[string]string
> > +
> > +	crawlExecutor  *mesos.ExecutorInfo
> > +	renderExecutor *mesos.ExecutorInfo
> > +
> > +	// this channel is closed when the program receives an interrupt,
> > +	// signalling that the program should shut down.
> > +	shutdown chan interface{}
> > +	// this channel is closed when there are no more outstanding tasks,
> > +	// and that the scheduler does not intend to receive more tasks
> > +	done chan interface{}
>
> I guess the way I describe them isn't very clear. So shutdown is closed
> when we receive an interrupt, signaling that the framework *should* shut
> down. When the scheduler sees that shutdown has been closed, it starts
> cleaning up the remaining tasks. Once that's done, it closes done.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/RENDLER/pull/29/files#r34116903>.
>